### ISSUE TITLE
Add Backup and Restore for Keyboard Settings

### ISSUE THREAD

**Issue by Nordlicht-13** [futo-org/android-keyboard#643]
### Use Cases

It would be nice to backup and restore the settings of the keyboard, so you could take them to a new phone, or restore them again after trying a nightly build.

**Comment by NicholasFlamy** [comment]
Similar to #125 but that feature includes sync which is not necessary for a keyboard, but it is nice to have. (I would actually use the Personal Dictionary if that would sync between my phone and tablet and if I could back it up and syncing my typing suggestions would be nice.)

This feature is necessary though, because once I set up my keyboard, I want to be able to restore that on a new device, or my tablet, etc.

**Comment by dean-futo** [comment]
Indeed, very similar to #125 - merging that issue into this one, to keep the focus on adding backup.

**Comment by favdish** [comment]
I am keeper of a zoo of Android phones/tablets and I definitely would need some form of **local** import/export, especially for the personal dictionary.
OTOH I certainly don't want any option or code that stores things in "the cloud".

**Comment by danielrosehill** [comment]
Thanks for saving me opening a discussion about this.

Just wanted to say that of all the backup settings, the one that might be very useful would be the personal dictionary in particular.

Consider an example in which a user develops quite an elaborate list of words over time, the user would definitely want some easy way to export that data so that if the device were lost, they wouldn't lose all their custom words.

Given the privacy-centric focus of the app, perhaps this could be implemented as a simple TXT or JSON dump onto a local path in the Android directory giving the user the ability to put the data wherever they wish.

**Comment by PhillyMay** [comment]
Indeed, a backup feature would be very much appreciated. 

Maybe it could be implemented as both an automatic and a Manual one.

The manual one could give you the Option to Export all settings, or just the settings/the wordbook etc.

The automatic backup feature could be scheduled, so it can be triggered by adding something to a wordbook vor periodically after x amount oft time (days, weeks, month)

**Comment by NicholasFlamy** [comment]
> OTOH I certainly don't want any option or code that stores things in "the cloud".

I think the only way FUTO would ever implement such a thing is if it syncs to your own self-hosted server.

**Comment by Nordlicht-13** [comment]
I was thinking of a backup-file on the phone, that you can backup for yourself and restore the settings, if you install the keyboard on an other or new phone.

**Comment by NicholasFlamy** [comment]
> I was thinking of a backup-file on the phone, that you can backup for yourself and restore the settings, if you install the keyboard on an other or new phone.

I think that should be implemented whether or not self-hosted sync is implemented. Both features are gonna require some kind of backend settings import/backup/sync changes and are related in implementation.

**Comment by gnattishness** [comment]
Config import/export is also useful for setups involving multiple user profiles like Graphene OS.
There you might want to install the keyboard on multiple profiles and have consistent settings.

Manual export is fine for my purposes.

**Comment by abb128** [comment]
This has been implemented in nightly

**Comment by NicholasFlamy** [comment]
> This has been implemented in nightly

Is there a build pipeline that releases nightly builds? (So that I wouldn't have to build it myself to test.)

**Comment by abb128** [comment]
@NicholasFlamy https://docs.keyboard.futo.org/improvements/nightly

**Comment by Nordlicht-13** [comment]
Obtanium Download: 
https://apps.obtainium.imranr.dev/redirect?r=obtainium://app/%7B"id"%3A"org.futo.inputmethod.latin.unstable"%2C"url"%3A"https%3A%2F%2Fdl.keyboard.futo.org%2Fnightly_version"%2C"author"%3A"dl.keyboard.futo.org"%2C"name"%3A"FUTO%20Keyboard%20%5BDev%20Build%5D"%2C"preferredApkIndex"%3A0%2C"additionalSettings"%3A"%7B%5C"intermediateLink%5C"%3A%5B%5D%2C%5C"customLinkFilterRegex%5C"%3A%5C"%5C"%2C%5C"filterByLinkText%5C"%3Afalse%2C%5C"skipSort%5C"%3Afalse%2C%5C"reverseSort%5C"%3Afalse%2C%5C"sortByLastLinkSegment%5C"%3Afalse%2C%5C"versionExtractWholePage%5C"%3Atrue%2C%5C"requestHeader%5C"%3A%5B%7B%5C"requestHeader%5C"%3A%5C"User-Agent%3A%20Mozilla%2F5.0%20(Linux%3B%20Android%2010%3B%20K)%20AppleWebKit%2F537.36%20(KHTML%2C%20like%20Gecko)%20Chrome%2F114.0.0.0%20Mobile%20Safari%2F537.36%5C"%7D%5D%2C%5C"defaultPseudoVersioningMethod%5C"%3A%5C"APKLinkHash%5C"%2C%5C"trackOnly%5C"%3Afalse%2C%5C"versionExtractionRegEx%5C"%3A%5C"%5C%5C%5C%5Cd%5C%5C%5C%5C.%5B%5E%5C%5C%5C%5C%5C%5C%5C%5C%5D%2B%5C"%2C%5C"matchGroupToUse%5C"%3A%5C"%240%5C"%2C%5C"versionDetection%5C"%3Afalse%2C%5C"useVersionCodeAsOSVersion%5C"%3Afalse%2C%5C"apkFilterRegEx%5C"%3A%5C"%5C"%2C%5C"invertAPKFilter%5C"%3Afalse%2C%5C"autoApkFilterByArch%5C"%3Atrue%2C%5C"appName%5C"%3A%5C"%5C"%2C%5C"shizukuPretendToBeGooglePlay%5C"%3Afalse%2C%5C"allowInsecure%5C"%3Afalse%2C%5C"exemptFromBackgroundUpdates%5C"%3Afalse%2C%5C"skipUpdateNotifications%5C"%3Afalse%2C%5C"about%5C"%3A%5C"%5C"%2C%5C"refreshBeforeDownload%5C"%3Afalse%2C%5C"supportFixedAPKURL%5C"%3Afalse%7D"%2C"overrideSource"%3A"HTML"%7D

**Comment by NicholasFlamy** [comment]
> [@NicholasFlamy](https://github.com/NicholasFlamy) https://docs.keyboard.futo.org/improvements/nightly

Thanks



> Obtanium Download: [link](https://apps.obtainium.imranr.dev/redirect?r=obtainium://app/%7B%22id%22%3A%22org.futo.inputmethod.latin.unstable%22%2C%22url%22%3A%22https%3A%2F%2Fdl.keyboard.futo.org%2Fnightly_version%22%2C%22author%22%3A%22dl.keyboard.futo.org%22%2C%22name%22%3A%22FUTO%20Keyboard%20%5BDev%20Build%5D%22%2C%22preferredApkIndex%22%3A0%2C%22additionalSettings%22%3A%22%7B%5C%22intermediateLink%5C%22%3A%5B%5D%2C%5C%22customLinkFilterRegex%5C%22%3A%5C%22%5C%22%2C%5C%22filterByLinkText%5C%22%3Afalse%2C%5C%22skipSort%5C%22%3Afalse%2C%5C%22reverseSort%5C%22%3Afalse%2C%5C%22sortByLastLinkSegment%5C%22%3Afalse%2C%5C%22versionExtractWholePage%5C%22%3Atrue%2C%5C%22requestHeader%5C%22%3A%5B%7B%5C%22requestHeader%5C%22%3A%5C%22User-Agent%3A%20Mozilla%2F5.0%20(Linux%3B%20Android%2010%3B%20K)%20AppleWebKit%2F537.36%20(KHTML%2C%20like%20Gecko)%20Chrome%2F114.0.0.0%20Mobile%20Safari%2F537.36%5C%22%7D%5D%2C%5C%22defaultPseudoVersioningMethod%5C%22%3A%5C%22APKLinkHash%5C%22%2C%5C%22trackOnly%5C%22%3Afalse%2C%5C%22versionExtractionRegEx%5C%22%3A%5C%22%5C%5C%5C%5Cd%5C%5C%5C%5C.%5B%5E%5C%5C%5C%5C%5C%5C%5C%5C%5D%2B%5C%22%2C%5C%22matchGroupToUse%5C%22%3A%5C%22%240%5C%22%2C%5C%22versionDetection%5C%22%3Afalse%2C%5C%22useVersionCodeAsOSVersion%5C%22%3Afalse%2C%5C%22apkFilterRegEx%5C%22%3A%5C%22%5C%22%2C%5C%22invertAPKFilter%5C%22%3Afalse%2C%5C%22autoApkFilterByArch%5C%22%3Atrue%2C%5C%22appName%5C%22%3A%5C%22%5C%22%2C%5C%22shizukuPretendToBeGooglePlay%5C%22%3Afalse%2C%5C%22allowInsecure%5C%22%3Afalse%2C%5C%22exemptFromBackgroundUpdates%5C%22%3Afalse%2C%5C%22skipUpdateNotifications%5C%22%3Afalse%2C%5C%22about%5C%22%3A%5C%22%5C%22%2C%5C%22refreshBeforeDownload%5C%22%3Afalse%2C%5C%22supportFixedAPKURL%5C%22%3Afalse%7D%22%2C%22overrideSource%22%3A%22HTML%22%7D)

@Nordlicht-13 nightly builds I only ever use for testing so that's the main thing I don't want auto updating lol.
